### ISSUE TITLE
Save an "interrupted" snapshot when backup is interrupted by SIGINT

### ISSUE THREAD

**Issue by aawsome** [restic/restic#2960]
Output of `restic version`
--------------------------
restic 0.10.0 (v0.10.0-23-g506fcfd4-dirty) compiled with go1.14.7 on linux/amd64

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------

When `restic backup` is interrupted by SIGINT, it does not save any snapshot. The only thing that can be used in following backups is the blobs that have been saved and included in an index file that also has been already saved.

This has many drawbacks:

- Users may wonder why snapshots are missing (e.g. for automated backup runs)
- The data already backuped is not accessible (can be only accessed after a `recover` run)
- pack files may have been written which didn't make it in an written index and are effectively "lost"
- a `prune` run will completely remove the interrupted backup which may not be what users wish.

IMO, we can do better! I propose to do the following after getting an SIGINT:

- immediately stop saving of data blobs
- BUT continue saving trees (which should now finish rather quickly)
- finish the index and save it
- also save the snapshot, but with added tag "interrupted"
- if everything of the above does not finish within a given time - e.g. 15s - cancel everything to let restic terminate.

Of course, this approach also involves ignoring snapshots with the tag "interrupted" when search for the parent snapshot.

While I already have a draft implementation which does exactly the steps I pointed out, I would like to use this issue to discuss the solution:

- Is the general way to follow ok?
- Are there some pitfalls I didn't see?
- Should we make the tag name and/or time to hard finish configurable for users?

What are you trying to do? What problem would this solve?
---------------------------------------------------------

See the problems above.

A user in the forum tried to find a workaround and didn't really suceed, see:
https://forum.restic.net/t/forget-instead-of-rebuild-index-then-recover-then-prune/3088

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------

Not today so far, but makes me happy on several other days :wink:

**Comment by buschjost** [comment]
I really like this idea. It would have helped me when I started using restic. 

For me 15 second sounds too short to be useful with any backend except local repositories. 

One option could be a timeout that is dependent on the backend (but rclone could be any backend...).

I am wondering if this could be a little more interactive and tell the user what is going on. For example (only when a local tty is detected?) show something like:

> Shutdown request received. Presevering intermediate data. Can take up to XYZ. 
>
> To immediately exit send signal again or press CTRL-C. 

And in case the signal is sent again or on keystroke it would immediately exit. With this the timeout could be longer, the user would know what is going on and has the choice to exit directly.

**Comment by jim-collier** [comment]
Agreed this is a great idea. But what about going farther - auto-saving the index in a draft state every so often? I know I may be (probably) using the wrong terminology here. 

Borg backup, for example, auto-saves a checkpoint archive every 30 minutes or so. That way even in the event of power failure or hard crash (i.e. no sigint), still at most you only lose 30 minutes of backup.

Either way, this feature would still be an improvement!

**Comment by ProactiveServices** [comment]
I'll be devil's advocate on this one as my gut reaction is negative. No offence intended!

Adding arbitrary tags to snapshots will foul up `forget`s. The user now has to be aware that their forget runs will have to take into account these tags, and has to then adjust their usage (manual or automated). Tags are no longer fair game; users must make sure none of the tags they used yesterday are now arbitrarily used by restic and the tags they use tomorrow cannot, either. If the tag names are not arbitrary we now have more complex configuration for gain in corner-cases which may not be useful for many users. Even if I do get to choose the tag name, I have to be completely sure that I have never used it, and I have to remember not to use it myself again.

We also now cannot simply grab a list of snapshot IDs and know that they contain all of the data we expect them to. When we pick out a snapshot to restore or mount, we must remember to check the tags.

When a backup is cancelled on purpose say, because the user notices unwanted data being backed up, the user has to (remember to, and...) delete the snapshot as well as the other necessities.

When I want to cancel restic I now have to wait up to 15s for it to finish cancelling, or send SIGINT twice, or any other new steps required. Previously it was one simple signal and I knew restic will briefly clean up then exit.

How is the exit code handled now? Data was backed up, a snapshot was added. But the backup did not complete.

You list a series of disadvantages of the current behaviour. I will refute them all by saying that no reasonable user can, upon experiencing a cancelled/failed backup process, expect there to be any useable data. They must expect any "unwanted" data to be removed by a prune. The fact that restic currently allows the uploaded data to be reintegrated is a great bonus.

**Comment by jim-collier** [comment]
@ProactiveServices 

That wasn't directed at me and I can't really comment on much of it - except this:

> no reasonable user can, upon experiencing a cancelled/failed backup process, expect there to be any usable data

I'm not sure what perspective you're coming at this from. Maybe from some point of view I can't currently wrap my head around, this isn't a completely insane perspective. Either way, I couldn't disagree more strongly with it, from my perspective. In fact, I'd say about any "reasonable user", exactly the opposite.

To wit:

- Most enterprise-grade backup software (whether commercial of FLOSS) can be interrupted for any reason at any time, and only lose at most some amount of data since a last checkpoint of some sort, with varying degrees of automatic recovery (typically fully automatic). For backing up enterprise data or large personal data sets, I can assure you this is typically an "assumed" feature.
    - For example, as the final decider at my employer for backup software, this was an explicitly hard requirement. (We were not content with just "assuming" its existence, although few serious contenders lacked it.) We have evaluated over a dozen products, and there is only one that loses all data on failure. For my own personal use, I evaluated at least a dozen more, including stuff that strays from strictly "backup" software. (E.g. looked into syncing software. All support seamless resumption after failure.) 
        - Crashplan, for example, met that requirement and many others. (But has since degraded so severely in features and performance over the years that we are now forced to change.) Recovery from any kind of hard crash or poweroff is completely transparent, which has happened uncountable times. Restores have never failed. (Just very slow.)
        - [Borg backup](https://github.com/borgbackup/borg) - arguably the most similar FLOSS competitor to Restic in terms of features (and even CLI interface) - [handles interrupted/aborted backups with ease](https://borgbackup.readthedocs.io/en/stable/faq.html#checkpoints-parts). It takes intermediate checkpoints every 30 minutes by default, and recovery/resumption is more or less automatic and transparent.
        - [Duplicacy](https://duplicacy.com/) is the only major dedicated cloud-capable backup product - either commercial or FLOSS (in this case both) - that we came across in our research, that cannot recover from an interrupted backup. In spite of a slate of impressive features, **that alone single-handedly eliminated it from the running**. There's simply no point in investigating the product any further.
        - The only reason Restic was considered for personal use, and is being considered for business use, is *specifically because* interrupted backups can be recovered. I believe you can find exactly that question (which wasn't super clear in the documentation) as one of my first interactions with the Restic community, long before I started using it. Granted, in practice it's a little on the painful side which I've discovered in real-world practice - which ranks it lower on our list for business - but can be done.
- There are innumerable, unpredictable, and non-enumerable reasons backups can be interrupted. A few examples:
    - Days-long "PSP" region-wide power shutdowns (in the SF bay area), straining the biggest battery backups and even gennys, which have affected even enterprise datacenters.
    - OS crashes.
    - Backup software crashes.
    - Accidental poweroffs/reboots.
    - Rolling blackouts during summer peak power usage.
    - Extended storm-related outages.
    - Extended internet outages.
    - Hardware failure, even affecting (less frequently) systems with high component redundancy.
    - Automatic reboots for patches. (It happens occasionally even to the most carefully-configured and well-managed systems.)
- Even over gigabit internet, a meager 15 TiB can take 20 days in a best-case scenario to back up to backblaze, for example. (The math may be optimistic. About 10 MB/s seems to be about the sustained max throughput for a single TCP connection to most cloud providers, and although Restic supports defining multiple connections, it appears from reading through the issues, that the pipeline as a whole is effectively limited to 1. Either way, practical experience suggests about double that time, with no resource bottlenecks, presumably just waiting for ACKs.)
- Mitigation strategies: Yes there are kludges to work around the limitations and potential disasters of a software product that might be inexplicably literally designed to not handle any one of the common interruptions listed above. For example:
    - Define multiple, smaller backup sets. Although Restic doesn't support this directly, with a fair amount of scripting, exhaustive testing, and trial-and-error, the notion of "backup sets" can be implemented. (Though dedicating a team to do that isn't necessarily on the "todo" list of most "reasonabl" users or organizations.) Backup sets could be partitioned by date range, file type, etc.
        - (Except in cases where that can't be done, or reasonably so.)
        - Either way, such fine-grained partitioning would require a custom solution to generate such sets, and would rely on the `--files-from` feature, which is currently "broken" in terms of how pretty much every other software utility in existence treats the contents of a `--files-from` file. (E.g. as a literal list of exact files. Not as globbing specifications.) In fact I'm developing such a utility as we speak to do that, including configurable maximum total set sizes. Will work with any product or utility that supports a `--files-from` feature, e.g. Restic, Borg, rsync, rclone... pretty much anything that primarily operates on a large number of files. (Restic is one of the few that will require a custom-filtered output in order to deal with it treating the list as glob specs.)
    - Data could be manually partitioned into smaller, similar-sized directories.
        - (Except in cases where it can't be, or can't reasonably be.)

In short, any software that could wind up costing a company hundreds or thousands of dollars - and much worse, put important data at risk - because it was *specifically "designed" to lose potentially vast amounts of data in the case of one of those common enough reasons above* (perhaps because of grossly misguided notions of what is expected of a "reasonable user"), is simply unacceptable, and a complete non-starter. Not just for me. Although I haven't polled my professional colleagues and cohorts about this, I've been around "a long time" and it would be an insane notion that any of them to accept such an intentional design flaw.

But then, maybe we just aren't "reasonable". ;-)

**Comment by greatroar** [comment]
I'm not sure about the proposed solution. Why not periodically write temporary shapshots with some tag (say `__restic_temporary__`), then promote the last one by removing the tag once the backup is done? That catches failures that are not signaled by a SIGINT, say a kernel panic.

**Comment by rawtaz** [comment]
@jim-collier You wrote a lot of text there. Just a few comments:

Without addressing each of the softwares you mention having support for interrupted backups, I took a look at https://borgbackup.readthedocs.io/en/stable/faq.html#checkpoints-parts since Borg is as you say one of the more common other backup software used. Reading that it seems to me that it will effectively work the same way for the user - an interrupted backup means that the already uploaded data is in the repository, and will be re-used when the user runs starts the backup run again. So how is this any different from what we already have in restic?

> In short, any software that could wind up costing a company hundreds or thousands of dollars - and much worse, put important data at risk - because it was specifically "designed" to lose potentially vast amounts of data in the case of one of those common enough reasons above (perhaps because of grossly misguided notions of what is expected of a "reasonable user"), is simply unacceptable, and a complete non-starter.

Are you referring to restic? I think anyone here understands that restic wasn't "specifically 'designed' to lose potentially vast amounts of data in the case of one of those common enough reasons above". Quite the opposite, it's designed such that already uploaded data can be re-used, in other words so that restarting the backup run will continue from where it left off (more or less). This is quite the opposite of what you seem to be saying.

---

@aawsome I'm not very keen on having a set or flexible timeout. A proper solution should not rely on having to "guess" (even if it's automatically adjusted) a timeout to let some process finish. In part because that's an ugly design and in part because there's never any guarantee that there will be time for that anyway. Whatever solution is made to solve the use case at hand, it should IMO be of the type where when the SIGINT or similar comes in, restic can terminate right away. So presumably this means some kind of intermediate restore points.

**Comment by jim-collier** [comment]
@rawtaz 

> an interrupted backup means that the already uploaded data is in the repository, and will be re-used when the user runs starts the backup run again. So how is this any different from what we already have in restic?

Not much different. Although in practice, Borg is currently more automatic - it automatically saves periodic snapshots, and automatically "recovers with them". (In quotes because that's pretty hand-wavy and since I'm not a contributor I can't explain it better except my own user experience.) Restic is, frankly, a little painful to recover from an interrupted backup, at least if you want to explicitly do so just in case. (I've done it three times now - once unnecessarily "by accident", twice due to interrupted backups. It was painful, took about 5 days, but was a good thing because both times found errors.)

But that's not my point. Yes it would be good if restic were more automatic about it, but my point wasn't to complain about the way restic works now, the main point of all that was to emphatically refute @ProactiveServices 's assertion that (my paraphrase) "no reasonable user should expect an interrupted backup to contain any useful data whatsoever".

> Are you referring to restic? I think anyone here understands that restic wasn't "specifically 'designed' to lose potentially vast amounts of data in the case of one of those common enough reasons above". Quite the opposite, it's designed such that already uploaded data can be re-used, in other words so that restarting the backup run will continue from where it left off (more or less). This is quite the opposite of what you seem to be saying.

That was also to rebut @ProactiveServices 's assertion. By "specifically designed" to lose data, I was referring to the fact that restic would have to go *backwards* in terms of features, in order to have no recovery mechanism from aborted backups. And *any* modern backup solution should be considered "broken" or at least not "feature complete", unless it has a way to recover the time, energy consumption, greenhouse-gas causing, and mission-critical data that may have been spent days or weeks working on before being interrupted.

**Comment by rawtaz** [comment]
Can someone explain to me how this is such a huge problem (if it is, in your opinion)? I mean, if a backup run is interrupted, then just re-run the backup job and let it complete? And regarding automated backups, you are *obviously* expected to keep an eye on those and notice if they don't complete successfully.

The chances that your backed-up machine dies at the very same time that you happen to have the last backup run (or a few, if that's the case) interrupted, and before you manage to re-run the backup job, is just so small that I don't see how this is a big problem. But regardless, it doesn't hurt to have an easier way to recover the data that has been uploaded so far, even if it's unlikely that you ever need to. But an implementation of this has to be nicely done, not a hack of sorts.

It sure won't happen very often when you do the initial backup, as no sane person does an initial backup without keeping the source around until the backup is complete.

All in all, I think it's a very rarely occurring situation we're talking about here.

**Comment by jim-collier** [comment]
@rawtaz 

> Can someone explain to me how this is such a huge problem (if it is, in your opinion)? I mean, if a backup run is interrupted, then just re-run the backup job and let it complete?

This seems to be becoming a more contentious issue than it's worth. My [lengthy reply above](https://github.com/restic/restic/issues/2960#issuecomment-701093894) was intended to hopefully offer a reset of expectations [idealistically] - or at least a well-substantiated alternate perspective - on what to me is an insane notion that it's no big deal to just discard interrupted backups. 

To be clear: Restic _does not even do that, and no serious enterprise-capable backup solution does that_ - so I'm not sure why that seems to be such a "controversial" stance, or why we're even debating it.

I repeat: Restic doesn't even do what you are suggesting.

I don't think it is necessary to relitigate my argument. I think I laid it out as well as I can, in that reply, though I may be able to further address what _sounds like_ (but what do I know) some highly flawed working assumptions on your part. I realize that reply is a little long as github comments go, but if you genuinely want to understand the perspective, rather than state that you disagree (I can't tell which at this point), then I think you might better understand where I'm coming from, with a more careful reading. (Or if not, no harm. I would have skimmed over it too. The point is, _if_ you want to understand my perspective if even just to formulate a better rebuttal, then it's in there.

First let's dismiss the notion that backups are small and inherently disposable. There may be very real and quite considerable resources involved in creating a snapshot: expensive time, valuable CPU slices [possibly even charged per minute of], considerable power consumption from the client all the way up the physical infrastructure chain to cloud storage, resulting greenhouse gas emissions - not to mention data transfer costs, and data storage costs.

Maybe for you, delta backups are small. Good for you. But your needs aren't the needs of others who may use, or be evaluating, Restic. So ideally you should keep at least these points in mind:

1. Any backup job - and indeed the product itself - is only as good as its first run.
1. _You have absolutely no idea - nor can you appreciate - how big a users first run may be_. It could take days. It could easily take weeks.
    - For me, Crashplan took _2 years_ for its first run to complete, and that was back in its heyday when it worked well. After they migrated me to "Small Business", it started at _8 years_ estimated completion just to make up the parts they lost, and it kept growing from there.
    - >1 week starts bumping up against IT policies of mandatory downtime for security patches. If you try telling them that they can't reboot because the backup software will lose all it's progress, guess what swift decision is going to be made about backup software?
1. You don't know how slow a user's internet connection might be.
1. You don't know how reliable a user's uptime and/or internet connection may be, even if in a datacenter. (Especially if located in the SF bay area - don't know how much news you're aware of from the region over the last 18 months.)
1. You don't know how much or how little time a user may be able to dedicate to scripting, moving data, and/or otherwise generating subsets of data as ad-hoc "backup sets" (which Restic has no inherent notion of or support for, it should be noted).
1. Any evaluation of serious backup products, and/or a more casual user's first experience, will be with the very first backup. Which considering their backup size and internet speed, could easily take weeks.
1. If you are serious about your backups, and saving money on cloud storage, you will watch your backups progress with a "--verbose" flag. (Or alternately carefully inspect and filter the logs after the fact or in real time.) There's really no good substitute for just sitting at the console watching filenames stream by, at minimum as an initial smoke-test of your include and exclude rules. (Yes there are hacks and kludges to approximate what will happen beforehand. But unfortunately Restic has no `--dry-run` flag.)
    - Then when you see it dive into a gigantic forbidden directly and/or back up ginormous files you forgot to exclude, what do you do? Wait for days or weeks for the job to finish? Or CTRL+Break right then and there, fix your excludes, and resume - possibly with a `recover` beforehand?

Look. I'm old. I've been doing this since long before GUIs existed. You could probably accurately imagine where I'm at career-wise, in terms of determining what enterprise products we use. I've used well over a dozen backup products over those many years. Both at the enterprise level, and personal level. At a personal level, I have alot of data, and it doubles roughly every 18-36 months.

**There is absolutely no way I would _ever_ consider investing time or money into a backup product, that lost _all progress_ due to interruption.** You can dismiss that as one guy's opinion. Or you can argue it all you want. But it won't change that basic fact. Since I started this discussion, I've "casually" pinged a few friends and colleagues about this, whom I estimated would have an opinion. All around my age and experience. So far it has been unanimous: nobody would even consider such a product. And I got more than one derisive snicker.

So again: feel free to dismiss this perspective as irrelevant because it doesn't conform to your assumptions and expectations. But I promise you, it's not irrelevant or rare, and you'd do Restic and its community a great a disservice by dismissing it.

I'd even go so far as to say that I find it moderately alarming (as a personal user and evaluating for our corporate use) that a Restic contributor even _holds_ such an opinion.

Honestly, as promising as Restic is, and as professional and caring as most of the contributors seem to be, it's looking like Borg is the winner here. I've got two parallel backups going at home: Borg to borgbase.com, and Restic to b2. Identical mirrors on different hardware, and roughly identical include/excludes (inasmuch as possible with different include/exclude syntax). Borg has tolerated and recovered far more gracefully than Restic from a similar number of interruptions. That's hugely important. But honestly - and meaning no personal disrespect - your apparent attitude about it as a Restic contributor - an attitude that doesn't seem isolated - honestly, is potentially the more relevant issue to consider in this evaluation. (I'm not just backing up my personal data. I'm evaluating products for company use to get the hell off of Crashplan ASAP. Which, by the way, handles interruptions completely seamlessly and transparently, with zero user intervention. [Granted they are in complete control of the server side so that should not only be assumed, but required.])

**Comment by rawtaz** [comment]
@jim-collier Please.. Noone here, me included, is questioning yours or others' use cases, or trying to dismiss them. If I were, I would have written things quite differently than asking you in what way this is a problem, as part of trying to understand you better. We're all here for constructive purposes, and I fully agree and respect that my use case and others' use cases are different.

I have read in detail all you wrote, but I still don't understand (and fine, call me stupid or anything else if you wish, that's fine by me) for example the following:

> I repeat: Restic doesn't even do what you are suggesting.

Please bear with me here - What is it I am suggesting that restic isn't doing?

> Then when you see it dive into a gigantic forbidden directly and/or back up ginormous files you forgot to exclude, what do you do? Wait for days or weeks for the job to finish? Or CTRL+Break right then and there, fix your excludes, and resume - possibly with a recover beforehand?

What you would do is:

1. Cancel the currently running backup (e.g. Ctrl+C).
2. Add the directory you hadn't previously excluded to the exclude options in your restic command.
3. Run the backup command again (this time with the correct exclusions).

Restic will then scan the files to be backed up again, but pick up where it left off in terms of uploading the actual data you want to back up. There's no need to wait for the job to finish, you just cancel the job, fix the configuration, and re-run the job. It will not start the uploading from scratch or square one, it will continue uploading where it left off the last (interrupted) backup run.

The very same approach goes for when the backup was interrupted by other factors, such as broken connectivity or a sudden power loss - all you have to do is start the backup run with restic again, and it will scan the files, then continue to upload the data it had not yet uploaded.

Does this make sense? If it doesn't, I must be missing or misunderstanding something very much :) I wouldn't be surprised if that's the case, as there seems to be a big disconnect between how we view this. Cheers.

**Comment by rawtaz** [comment]
> Why not periodically write temporary shapshots with some tag (say __restic_temporary__), then promote the last one by removing the tag once the backup is done?

I'm thinking make it a snapshot with the special ID of 00000. Then restic can just replace that same ID over and over again when saving intermediates and it won't touch tags.

**Comment by jim-collier** [comment]
@rawtaz 

> call me stupid or anything else if you wish, that's fine by me

I wouldn't do that even if I believed it to be true. But 1) I would need substantially more data before coming to that conclusion, 2) There's probably three people ever I've felt qualified to make such a conclusion about - and I think they've all since died from stupidity - and 3) Even from limited data I'm pretty sure that's not the case here.

I think we're just talking past each other. And may in fact be in violent agreement. In fact, I feel perplexed by what "seems" like a complete 180 in your position, which is often a sign of talking past each other.

I think part of the problem was with these two quotes from you:

> I mean, if a backup run is interrupted, then just re-run the backup job and let it complete?

and

> no sane person does an initial backup without keeping the source around until the backup is complete.

Which sounds awfully similar to - and I read to be in support of - the comment from @ProactiveServices :

> no reasonable user can, upon experiencing a cancelled/failed backup process, expect there to be any useable data.

It's not that I assumed or jumped to conclusions, I read it as literally the same. Sounds like I read wrong, I'm happy to own that. And it does seem like I read it wrong, because you went on to say things like,

> Restic will then scan the files to be backed up again, but pick up where it left off in terms of uploading the actual data you want to back up. There's no need to wait for the job to finish, you just cancel the job, fix the configuration, and re-run the job. It will not start the uploading from scratch or square one, it will continue uploading where it left off the last (interrupted) backup run.

and 

> The very same approach goes for when the backup was interrupted by other factors, such as broken connectivity, a sudden power loss or just having to temporarily cancel it for some maintenance or patching - all you have to do is start the backup run with restic again, and it will scan the files, then continue to upload the data it had not yet uploaded.

I hope that it's at least been clear, that I am whole-heartedly in favor of those last two sentiments!

So I think we're good, for what it's worth!

I'm curious though - if those last two points are true (and I did not know it was that automagic):

1. How does Restic know to continue an aborted prior run, and where to pick up? (Obviously it's possible, I'm just curious at a high level. I read a great deal of the Restic documentation and design before going further, and don't recall that. But would love to - do you know off the top of your head where/what I could search for that, maybe what keywords?)

2. Why is this particular issue we're discussing even an open issue then, if Restic can already successfully resume from interruption?

**Comment by rawtaz** [comment]
> How does Restic know to continue an aborted prior run, and where to pick up? (Obviously it's possible, I'm just curious at a high level. I read a great deal of the Restic documentation and design before going further, and don't recall that. But would love to - do you know off the top of your head where/what I could search for that, maybe what keywords?)

I should probably let someone with more insight into the internals of restic answer that, but the short summary is: When backing up, restic [periodically writes index files](https://github.com/restic/restic/pull/310) to keep a record of the uploaded data. Even if there's no snapshot created in the end (due to interrupted backup), these indexes are stored in the database for the data that has been uploaded so far. Next time restic runs, it is then able to find the uploaded data through these indexes, and thereby make use of it again without having to upload it a second or more time. This effectively makes it continue from where it left off. There's naturally a lot more going on, but that's the bigger picture of it.

So it's less of "record an incomplete snapshot as you go" and more of a "automatically reuse already uploaded data" thing.

> Why is this particular issue we're discussing even an open issue then, if Restic can already successfully resume from interruption?

This issue is about a slightly different thing, namely the ability to *easily access the partially uploaded data*. So, if you have a backup that was interrupted, the author of this issue wants to enable the user to easily access the files that were (I presume, completely) uploaded and stored in the repository within the time before the backup was interrupted. This is not easy to do today, the closest to it is the `recover` command which personally I haven't used so I can't say much about it.

This is where my slight skepticism or need for better understanding of the use case comes in, as I don't really understand how there can be a great need for this feature. Granted, if you run a large initial backup which gets interrupted half way, and at the same time the data you are backing up is somehow lost, then indeed there's a need to access the partially uploaded data. At that time you can use the `recover` command albeit not perfect and easy to use. But I think the chance of ending up in this situation is so extremely small, and this is why I was asking to what extent this is an actual problem for people. Besides this in my opinion extremely rare use case I would expect that if your backup is interrupted, you simply re-run it and in the end you'll have a good snapshot and everything will be fine, hence my questions.

I hope this clears it up a bit, let us know otherwise! FWIW, in other settings I use the extremely well established and enterprisey backup software named Veeam Backup & Recovery, and it pretty much works the same as restic in terms of resuming interrupted backups (it will not upload already uploaded data anew, but you can't access the already uploaded but "interrupted" data through a restore point), so I was a bit confused by what you wrote earlier :)

**Comment by MichaelEischer** [comment]
restic uploads an intermediate index at least every 10 minutes, which lists all file chunks uploaded in the meantime. A later backup run then uses that information to avoid reuploading these file chunks which are already listed in the index. The checkpoints in borg seem to provide a similar feature. What is not clear to me from the linked borg documentation (https://borgbackup.readthedocs.io/en/stable/faq.html#checkpoints-parts) is whether subsequent backup runs in borg will be able to use these checkpoints to skip files as "unchanged" and therefore also substantially speed up the file system scan. That latter feature is what would be necessary to quickly resume an interrupted multi TB backup. restic normally uses the latest snapshot for a backup set as reference point when deciding which files have to be processed again as these have changed and which files can be skipped as these are unchanged. For the initial backup that's (currently) not possible and there an interrupted snapshot would help a lot, as restic could just skip the already processed files when continuing.

> Restic is, frankly, a little painful to recover from an interrupted backup, at least if you want to explicitly do so just in case. (I've done it three times now - once unnecessarily "by accident", twice due to interrupted backups. It was painful, took about 5 days, but was a good thing because both times found errors.)

That sounds like you ran `rebuild-index` after an interrupted backup, which shouldn't be necessary (unless the repository was damaged or lost files on the repository storage side)? If you have to do more than just calling `unlock` then you've hit a bug. The idea is that restic just requires the repository storage to not loose or damage _successfully_ written data. A command as a whole only completes successfully when the connection to the repository storage works, an interrupted connection is assumed to be possible at any time. In this case the repository should still be in a usable state (but may contain some garbage data) and the interrupted operation has to be started from scratch. That said, I'm currently aware of a few corner-cases where manual intervention could become necessary (e.g. delete an incomplete file), these will eventually be fixed to be handled automatically but that will still take quite some time. In case files in the repository are corrupted (or the client uploaded damaged data due to a bit-flip), then its usually still possible to access not damaged parts of the backup (this is currently only partially tested and may require manual recovery steps). However, interruptions are (currently) assumed to be not too frequent such that it is possible to successfully complete operations most of the time.

> Even over gigabit internet, a meager 15 TiB can take 20 days in a best-case scenario to back up to backblaze, for example. (The math may be optimistic. About 10 MB/s seems to be about the sustained max throughput for a single TCP connection to most cloud providers, and although Restic supports defining multiple connections, it appears from reading through the issues, that the pipeline as a whole is effectively limited to 1. Either way, practical experience suggests about double that time, with no resource bottlenecks, presumably just waiting for ACKs.)

The file chunking and deduplication part should, given enough IO bandwidth and CPU cores, for large files (10+MB, or rather not small files) be able to process hundreds MB/s (maybe even 1GB/s but that's a quite rough estimate) the limitation right now is the upload to the storage backend. And yes I'm aware that even at 1GB/s it would take 4 hours just to fully rescan a 15TB dataset after the initial backup operation was interrupted. For small files the pipeline concurrency is effectively limited to 1 for high-latency backend (more than a few dozen milliseconds), this will be fixed in #2696.

@jim-collier `a meager 15 TiB` probably highlights the _major_ difference of expectations between you and @rawtaz . A 15 TB repository, from what I can tell, is already at the upper end of the _typical_ restic backup repository sizes. Just for the sake of argument: A 1TB backup data set which can be read at 125MB/s from a HDD just takes 2 hours for a full rescan in case the backup is interrupted. Although this duration might be annoying, it's not much of a problem (at least I'd assume that electricity and other required resources are available for a few hours without interruptions). Increase the backup size by one order of magnitude and then we're at roughly a day, at which point interruptions definitely start to become a problem. A multi TB backup delta (either initial backup or a later one) is currently outside the size range where I would call the usability of restic good. For a repository which requires years for the initial upload it only gets worse.

I'd like to add another reason for keeping the resource usage low: It's nice when the CPU usage (of restic) stays low enough during a backup such that the notebook fans don't have to start spinning.

Faster resumption of interrupted backups would definitely be a nice feature; intermediate snapshots on their own are probably rarely enough useful to justify cluttering the snapshots list with them and we already have a partial workaround for these using the `recover` command.

**Comment by aawsome** [comment]
First, thanks everybody so far for this intense discussion! :+1: 

While I would like to summarize the arguments and try to propose a good way to proceed with this topic, I actually miss a bit time, so I have to postpone this (and hope to see more arguments in the meantime here :wink:)

Regarding the idea of "saving intermediate snapshots" I just wanted to throw in two points:

1. Cleaning up those "intermediate snapshots" is an issue. If the `backup` run also cleans up those, backup needs to remove snapshots which contradicts the actual philosophy that `backup` only adds files to the repo and deletes nothing except lock files. Also then `backup` may no longer work in some settings like append-only repos. The other alternative is that restic *must* be able to handle lot's of present "intermediate" snapshots smoothly.
2. Regarding the implementation, I currently see no way to easily add this functionality without a heavy redesign of the `archiver` codebase... But maybe there is someone who has a good idea about this.

**Comment by jim-collier** [comment]
@rawtaz 

I think I get it now. These phrases cinched it:

> restic normally uses the latest snapshot for a backup set as reference point when deciding which files have to be processed again as these have changed and which files can be skipped as these are unchanged.

> restic uploads an intermediate index at least every 10 minutes, which lists all file chunks uploaded in the meantime. A later backup run then uses that information to avoid reuploading these file chunks which are already listed in the index.

> At that time you can use the `recover` command albeit not perfect and easy to use.

I think my major disconnect was a lack of this more nuanced understanding:

1. Restic uses an intermediate index internally, to not reupload the same data next time, and
1. That data is not available to users, until either:
    1. User runs the backup again to completion - in which case the already uploaded data won't need to be reuploaded; **or**
    1. User runs `recover`. (In extreme cases sometimes also requiring a `rebuild-index` before that, according to some past issue I believe @fd0 comment on with about as much).

Specifically, I wasn't aware of `2.i`. I had been under the impression that `2.ii` was the only option. That's a _huge_ "improvement" in functionality! :-)

I was under the belief that Borg treats its intermediate checkpoints as accessible to the user without further action. But now that I think about it, I don't know how I came to that understanding. So it could just be speculation masquerading as belief.

And yes, I had been running `rebuild-index` before `recover`. Super painful. Took several days, like `prune` (which I suppose should be expected). 

At least now, I know that `recover` isn't necessary to recover interrupted backups - just run it again and let it finish. Thank you!

**Comment by jim-collier** [comment]
@MichaelEischer 

> restic uploads an intermediate index at least every 10 minutes, which lists all file chunks uploaded in the meantime. A later backup run then uses that information to avoid reuploading these file chunks which are already listed in the index. 

It's possible the documentation is already this clear, and it just didn't register when I read it the first few times. (Not sarcasm.) But that blurb really drives the point home, thanks. I think what could probably be more clear, is this:

_In the event of interruption, the only way for users access the already uploaded data is 1) run the backup again and let it finish [no matter how many interruptions that takes], or 2) run `restic recover`. With first option being the preferred and intended method._

> The checkpoints in borg seem to provide a similar feature. What is not clear to me from the linked borg documentation (https://borgbackup.readthedocs.io/en/stable/faq.html#checkpoints-parts) is whether subsequent backup runs in borg will be able to use these checkpoints to skip files as "unchanged" and therefore also substantially speed up the file system scan.

From my experience with just that scenario, yes. What I'm not sure about, is if the user can directly access those intermediate snapshots. I believed so, but since I can't recall how I "know" that, now I'm not sure! (I could read the docs again but... I feel my countless hours of reading are finally at an end, time to just use the stuff and get s--t backed up.)

> The file chunking and deduplication part should, given enough IO bandwidth and CPU cores, for large files (10+MB, or rather not small files) be able to process hundreds MB/s (maybe even 1GB/s but that's a quite rough estimate) the limitation right now is the upload to the storage backend.

**Throughput to Backblaze b2**

This is a side discussion since it came up. I just want to be clear that I'm not complaining about it, or asking for help. I'm only breaking it down here, because it has come up often across a few of the issues I've opened or commented on in recent months. For some reason people think I have a problem with speed to backblaze b2. Either in terms of personal attitude, or actual performance. Neither is true. Sure, Restic could be improved to better leverage multithreading somewhere in the pipeline, in turn to better leverage multiple TCP connections to b2. But for now, I'm fine with the actual backup throughput. And I'm also aware that prunes will be much faster soon enough - so, all good!

But it might be worth breaking the performance down in more technical detail. Maybe this will help others in in future searches with concerns about their own backup throughput to Backblaze b2.

My server and 5x3-way mirror ZFS array can keep a local gigabit link saturated - at about 108 MB/s - without stress.

And while backing up, not a single hardware resource is anywhere near stressed. CPU cores low, RAM to spare, no pagefile activity, iotop mostly idle, internet link only about 10% of tested, verified real-world capacity - averaging about 10 MB/s, variable between about 5 and 15 MB/s. (All dependent on sample rate and reporting frequency of course.) Disks lights lazily blipping.

Most of restic's and the computer's time is literally spent waiting on TCP, specifically acks from b2.

Scanning, hashing, and compression don't get the chance to stress anything.

My internet connection is a dedicated gigabit fiber to the backbone. In speed tests to fast regional servers it sustains UDP at a steady 950 Mb/s both up and down. Between 5-10ms ping. Which by 2020 standards, is decent. And more importantly, those numbers are very stable.

But TCP of course is a different story, what with the [archaic traffic congestion algorithms and all](https://www.zdnet.com/article/fixing-the-unfairness-of-tcp-congestion-control/). (Thank god for [HTTS/3](https://thenewstack.io/http-3-replaces-tcp-with-udp-to-boost-network-speed-reliability/).)

Typically on a fast, low-latency connection, no matter the location, environment, or host hardware/OS involved, about that throughput (5-15 MB/s) seems to me, fairly typical for a **single** TCP connection to real-world cloud-based storage. Which is [more or less expected](https://cloud.google.com/solutions/tcp-optimization-for-network-performance-in-gcp-and-hybrid) for such a single TCP connection.

I have no problem with the speed of Restic that I'm observing. (Or more accurately, speed to b2 over a single TCP connection.) 

10 MB/s is a far cry better than Crashplan's 115 KB/s ([10 gb per day](https://support.code42.com/Administrator/Cloud/Troubleshooting/Backup_speed_does_not_match_available_bandwidth)).

And sufficient for my personal daily backups, after the initial backup. (And the initial backup did in fact _complete_, ever. That's not trivial. After Crashplan dropped their home product, my estimated completion time was 8 years, and kept growing as data was added faster than it could back up.)

(FWIW, Borg is similarly constrained at least to Borgbase.)

If Restic supported multiple TCP connections to b2 - meaning in actual effective use, not the `-o b2.connections` flag which seems to be ineffective in terms of affecting total throughput - then I imagine that would improve throughput to b2. I understand from several other issues on the matter, that the `-o b2.connections` flag is actually honored (which I can verify in htop), it just doesn't matter that much, since restic is only able to upload on one thread at a time. (Or something like that - I'm not fully up on - nor personally vested in - the issue, other than understanding it's a non-trivial effort to "fix".)

**How big is too big of a backup?**

> A 15 TB repository, from what I can tell, is already at the upper end of the typical restic backup repository sizes.

I have to take your word for what Restic is reasonably capable of dealing with.

But: 

_For whatever it's worth I would really like to encourage the Restic team to reconsider what a "large" backup is_!

15 TiB isn't that big in these days of digital cameras and video cameras - even just (unnecessarily) exploding megapixel phone cams - with storage requirements growing exponentially, and the increasing use of raw formats. (While camera megapixels may be leveling off due to diminishing returns involving quantum limits, shooting speeds keep rising, and camera storage is still growing exponentially in capacity and speed.)

Video file sizes are growing faster than ever. Even as compression algorithms get (only seemingly logarithmically) better, pixel dimensions are still growing at least geometrically, from 2k, to 4k, 8k... and soon enough (in the relevant context of development cycles), 12k. Meanwhile bit depth is increasing from 8bpc to 10 to 16 to even 32 for some intermediate file formats.

Hobby use of virtual machines is growing by leaps and bounds.

Meanwhile, _a single 16 TB drive on Amazon is just $360_.

20 TB of usable storage in a 3-way RAID-Z array would set you back only $600. (Granted single-parity raid is an increasingly terrible idea, even if ZFS design and top-down tree resilver mitigates the specific risk by a fair amount.)

Cloud storage is cheaper than ever. 20 TB of b2 storage runs just $100/month. (But even then, 20 TB of files to back up rarely translates to 20 TB of immediate cloud storage need. My personal data, for example - while not very compressible, is highly redundant.)

Granted, there is no Crashplan plan available at any price, that can reasonably deal with 20 TB of dynamic data. But I believe Restic should set it's sights on beyond that. If it falls short, at least you may get farther than you are now. (And I realize there's no well-defined "you" in most FLOSS.) But either way, "you" are going to face this law of exponential growth. 

To illustrate just to drive the point home more (or make the horse more dead), My own data doubles every 18-24 months-ish. Which is fairly standard for millions of people like me into digital media creation. In just the next two years, I'll generate more data than all the data I've generated between about 1986, and now. That's the norm, not the exception. No matter how little data you have now, it doesn't much matter: exponential growth catches up.

(I get that Restic isn't the only constraint, it's an ecosystem. But everything else - including CPUs, disk speeds, and throughput to say b2 - is going to keep up. Restic needs to keep up - e.g. prune faster, utilize concurrent TCP connections better, etc. I know that's all more or less in the works. Just reiterating the importance of not thinking "too small" when considering typical backup needs. Of course I'm aware that I can help. I might. I contribute to other projects in various ways. I'm actually learning Go now. But not everyone can contribute to every FLOSS product they use. Sometimes you just need to be OK consuming.)

**TL;DR**:

Shoot for the moon! That way if you miss, at least you'll be among the stars. ...Floating, with less than an hour of oxygen supply, until you eventually fall into a star or black hole - if there's anything left of your corpse by then after eons of bombardment by relativistic cosmic rays.

Phew.

**Comment by ProactiveServices** [comment]
@jim-collier This thread is getting OT for this issue; a great place for discussion is the [restic forums](https://forum.restic.net/). One of the various threads there discuss the [larger repo sizes](https://forum.restic.net/t/cern-is-testing-restic-for-their-backups/1523/16) that are known to be using restic, so best to acquaint yourself with those discussions before assuming what the software is working on and what the developers think about how their software is used.

**Comment by jim-collier** [comment]
@ProactiveServices I appreciate that you are a contributor, and that it turns out the member I was responding to who I thought was a contributor and suggested that 15 TB may be pushing it, apparently isn't a contributor in hindsight. That said, he's a much more pleasant fellow, doesn't hide behind passive-aggressive language, or lecture users that 

> no reasonable user can, upon experiencing a cancelled/failed backup process, expect there to be any usable data

But fair enough on OT and forums.

**Comment by rawtaz** [comment]
If you are referring to @MichaelEischer, I can assure you that he is a contributor. He's spending day and night making restic better, with razor sharp precision. One cannot contribute more than he does.

With that said, please let us all leave this discussion once and for all, and keep any further comments in this issue 100% highly on topic. Thanks everyone!

**Comment by jim-collier** [comment]
@rawtaz yep fair enough. I keep forgetting you have a forum, in spite of having engaged on it. :-/ That said, I was trying to offer (with some investment in time) what I hoped would be useful feedback and perspective to restic devs, from close to 40 years as a heavy personal and user and corporate decision-maker of backup products and services covering commercial, FLOSS, and home-spun solutions. It's certainly possible I'm overestimating that potential value and/or uniqueness, and wouldn't be the first time. ;-)

**Comment by greatroar** [comment]
Replying to @aawsome,

> Cleaning up those "intermediate snapshots" is an issue. If the backup run also cleans up those, backup needs to remove snapshots which contradicts the actual philosophy that backup only adds files to the repo and deletes nothing except lock files. Also then backup may no longer work in some settings like append-only repos. The other alternative is that restic must be able to handle lot's of present "intermediate" snapshots smoothly.

If backup removes something that it itself had added, I'd still say it only adds files, as long as it doesn't delete the temporary before the next (temp or final) snapshot is written. But append-only configs are an issue. The other problem with my suggested approach is that the intermediates produce garbage that needs to pruned away later.

> Regarding the implementation, I currently see no way to easily add this functionality without a heavy redesign of the archiver codebase... But maybe there is someone who has a good idea about this.

Me neither, I was shooting from the hip. I guess the archiver would have to keep track of a temporary top tree, and attach the trees that it completes to it. Of course, the tricky part is that the tree for `foo/bar` will be completed before `foo`, so it would have to make up some directories as it goes.

**Comment by jim-collier** [comment]
@greatroar 

> But append-only configs are an issue. The other problem with my suggested approach is that the intermediates produce garbage that needs to pruned away later.

If this helps. Understanding that Borg and Restic are very different, and that Borg can take advantage of a dedicated server which changes the calculus - and also appreciating that "not invented here" inevitably influences even the most abstract conceptual thinking (not always for the worse):  The way Borg handles it is that in append-only mode, it adds newer metadata that tells it to "ignore what I said previously". (Kind of like how we had add/edit/delete DVD modes, back in the day, in spite of dealing with a write-once-only medium.) IOW to the user, everything looks the same as if it hadn't been operating in write-only mode, as if stuff had been deleted. If Borg is later then run in read/write mode and pruned, it will make those changes, including deletes, permanent.

(Actually...your mentioning of attaching trees seems to indicate you might have already been on a similar logical track...)

**Comment by greatroar** [comment]
That's sort of what I was thinking, but the question remains where the required metadata would live. Tags are the obvious backward-compatible place for them, but the objections to invading tag space are valid. @rawtaz suggested using the magic all-zero id, but there's only one of those. Thinking one level down in the implementation, the snapshots/ directory in a repo could also house snapshot files with distinguishing names (say tmp-$id)...

**Comment by rawtaz** [comment]
> @rawtaz suggested using the magic all-zero id, but there's only one of those

I thought that we only need one (for this feature at least), as it would be replaced every time restic saves this "intermediate" snapshot. But yes, it will not work with append-only mode, and that's too breaking IMO, we have to keep supporting that mode.

**Comment by rawtaz** [comment]
I'm still not sure that there's an actually big enough problem to solve here. Restic already supports resuming backups. I'd argue that it's extremely rare that people need to restore files from an incomplete backup run. Do we have any actual data telling us that this is an actual problem we're discussing here?

**Comment by jim-collier** [comment]
> I'm still not sure that there's an actually big enough problem to solve here. Restic already supports resuming backups. I'd argue that it's extremely rare that people need to restore files from an incomplete backup run. Do we have any actual data telling us that this is an actual problem we're discussing here?

FWIW I'd agree with that sentiment, and I've had multiple interrupted backups. What had me freaked out (metaphorically speaking), was not _realizing that I didn't need to do a[n optional `repair-index`] and `recover` every time.  (The former as a "just in case", since, you know, critical data). When I just recently learned that backups would have been faster and faster after each interruption had I not even done anything at all (not having to upload already backed up stuff), until it finished, that would have been a game changer.

May I suggest that an easier alternative to this, might be making it even more abundantly clear in the docs and command-line help, these two options for dealing with an interrupted backup (the latter only being an edge case if you lose the source before finished).

Yes, I'm pretty confident that these facts could be pieced together from existing documentation. But if you consider that 1) I read ALL of the documentation thoroughly, many parts multiple times (as part of years-long review process) - in part knowing well in advance that I would have interrupted backups, 2) I in fact had multiple interrupted backups during initial backup, 3) I read the docs some more, 4) engaged on issues and forum, and 5) STILL wasn't aware that `restore` was unnecessary, just let the backup finish! Oh and also 6) if you believe that I'm _not_ below average Restic user intelligence [possibly a leap there]... then, going overboard on messaging about interrupted backups might either be A) a better solution, or B) worthwhile by itself either way.

**Comment by rawtaz** [comment]
Yes, we might add info about those commands and how to deal with an interrupted backup that you for some reason cannot restart, into an upcoming troubleshooting section. It's so rare that I don't think it needs to be made more prominent than such a location.

**Comment by jim-collier** [comment]
@ rawtaz

I _can_ restart backups, and now do. I just didn't know that was possible before. Earlier, I was operating under the misunderstanding that in order to restart a backup, you had to first `recover`.

**Comment by rawtaz** [comment]
@jim-collier I know, I understand that. I was just answering your suggestion about documenting when and how to use the two commands `rebuild-index` and `recover`.

**Comment by aawsome** [comment]
Just wanted to throw in that #3121 might give more opportunities to handle interrupted backups, e.g. using an "interrupted" snapshot *and* the standard parents as new parents.

So I'll wait for #3121 to finish and then come back to the discussion here.

**Comment by aawsome** [comment]
I worked a but on #3230 and think this is much more promising than thinking about saving "interrupted snapshots" or "intermediate snapshots" in fact it kind of implements both, but the "snapshot" (which is not a real one, but instead a list of local directories and corrseponding tree IDs) is not saved in the repo, but instead on the local disc. IMO this is enough to tackle the "speed-up" issue with aborted backups.

To solve the other issue - accessing files from an aborted backup, there is already `recover` and (if the source didn't change) the possibility to "resume" (i.e. rerun) the backup.

**Comment by patrickwolf** [comment]
> To solve the other issue - accessing files from an aborted backup, there is already `recover` and (if the source didn't change) the possibility to "resume" (i.e. rerun) the backup.

Actually recover doesn't seem to recover files left from a failed upload currently. Would be great if it did. Created a request for it #5287

**Comment by MichaelEischer** [comment]
With the refactored SIGINT handling in restic 0.17.x, it should now actually be possible to implement this is a half way clean way. (Just mentioning this in case someone wants to take a stab. Be warned that it remains a complex change nevertheless)
